diff --git a/third-party/fizz/src/fizz/crypto/signature/Signature.cpp b/third-party/fizz/src/fizz/crypto/signature/Signature.cpp
index f248957..de56992 100644
--- a/third-party/fizz/src/fizz/crypto/signature/Signature.cpp
+++ b/third-party/fizz/src/fizz/crypto/signature/Signature.cpp
@@ -149,8 +149,14 @@ void rsaPssVerify(
     throw std::runtime_error("Could not update verification");
   }
 
-  if (EVP_DigestVerifyFinal(mdCtx.get(), signature.data(), signature.size()) !=
-      1) {
+  if (
+    EVP_DigestVerifyFinal(
+      mdCtx.get(),
+      // const_cast<unsigned char*> is needed for OpenSSL 1.0.1 on Debian 8,
+      // which HHVM currently supports
+      const_cast<unsigned char*>(signature.data()),
+      signature.size()) != 1
+    ) {
     throw std::runtime_error("Signature verification failed");
   }
 }
