Index: hhvm-nightly-2019.09.23/third-party/folly/src/folly/io/IOBufQueue.h
===================================================================
--- hhvm-nightly-2019.09.23.orig/third-party/folly/src/folly/io/IOBufQueue.h
+++ hhvm-nightly-2019.09.23/third-party/folly/src/folly/io/IOBufQueue.h
@@ -53,14 +53,14 @@ class IOBufQueue {

     WritableRangeCacheData(WritableRangeCacheData&& other)
         : cachedRange(other.cachedRange), attached(other.attached) {
-      other.cachedRange = {};
+      other.cachedRange = {nullptr, nullptr};
       other.attached = false;
     }
     WritableRangeCacheData& operator=(WritableRangeCacheData&& other) {
       cachedRange = other.cachedRange;
       attached = other.attached;

-      other.cachedRange = {};
+      other.cachedRange = {nullptr, nullptr};
       other.attached = false;

       return *this;
