diff --git a/third-party/mcrouter/src/mcrouter/lib/network/gen/MemcacheRouterInfo.h b/third-party/mcrouter/src/mcrouter/lib/network/gen/MemcacheRouterInfo.h
index 908ca1e3..167e04fa 100644
--- a/third-party/mcrouter/src/mcrouter/lib/network/gen/MemcacheRouterInfo.h
+++ b/third-party/mcrouter/src/mcrouter/lib/network/gen/MemcacheRouterInfo.h
@@ -124,8 +124,6 @@ struct MemcacheRouterInfo {
 } // namespace memcache
 } // namespace facebook
 
-#include "mcrouter/lib/network/gen/MemcacheThriftTransport.h"
-
 #include <cstddef>
 #include <memory>
 #include <vector>
