diff --git a/third-party/mcrouter/src/mcrouter/routes/ShardSplitter.h b/third-party/mcrouter/src/mcrouter/routes/ShardSplitter.h
index 82676882..dd16712b 100644
--- a/third-party/mcrouter/src/mcrouter/routes/ShardSplitter.h
+++ b/third-party/mcrouter/src/mcrouter/routes/ShardSplitter.h
@@ -8,14 +8,11 @@
 
 #include <chrono>
 
+#include <folly/dynamic.h>
 #include <folly/Range.h>
 
 #include "mcrouter/config.h"
 
-namespace folly {
-struct dynamic;
-} // namespace folly
-
 namespace facebook {
 namespace memcache {
 namespace mcrouter {
