diff --git a/third-party/thrift/src/thrift/lib/cpp2/transport/http2/client/H2ClientConnection.cpp b/third-party/thrift/src/thrift/lib/cpp2/transport/http2/client/H2ClientConnection.cpp
index afdc607..318ae84 100644
--- a/third-party/thrift/src/thrift/lib/cpp2/transport/http2/client/H2ClientConnection.cpp
+++ b/third-party/thrift/src/thrift/lib/cpp2/transport/http2/client/H2ClientConnection.cpp
@@ -82,7 +82,7 @@ H2ClientConnection::~H2ClientConnection() {
 std::shared_ptr<ThriftChannelIf> H2ClientConnection::getChannel() {
   DCHECK(evb_ && evb_->isInEventBaseThread());
   return std::make_shared<SingleRpcChannel>(
-      *evb_, [this](auto* self) { return newTransaction(self); });
+      *evb_, [this](auto* self) { return this->newTransaction(self); });
 }
 
 void H2ClientConnection::setMaxPendingRequests(uint32_t num) {
