diff --git a/third-party/webscalesqlclient/src/CMakeLists.txt b/third-party/webscalesqlclient/src/CMakeLists.txt
index 6eede6adb4..f2b2202d7e 100644
--- a/third-party/webscalesqlclient/src/CMakeLists.txt
+++ b/third-party/webscalesqlclient/src/CMakeLists.txt
@@ -175,7 +175,6 @@ INCLUDE(cpu_info)
 INCLUDE(zlib)
 INCLUDE(libevent)
 INCLUDE(ssl)
-INCLUDE(readline)
 INCLUDE(mysql_version)
 INCLUDE(libutils)
 INCLUDE(dtrace)
@@ -475,8 +474,6 @@ INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/include)
 MYSQL_CHECK_ZLIB_WITH_COMPRESS()
 # Add bundled yassl/taocrypt or system openssl.
 MYSQL_CHECK_SSL()
-# Find system readline.
-MYSQL_CHECK_READLINE()
 # Add libevent
 MYSQL_CHECK_LIBEVENT()
 
