--- bundled_fbmysqlclient-orig/CMakeLists.txt	2021-07-09 15:48:32.000000000 -0500
+++ bundled_fbmysqlclient/CMakeLists.txt	2021-07-09 16:08:16.000000000 -0500
@@ -559,6 +559,11 @@
   ADD_DEFINITIONS(-DEXPORT_SYMVER16)
 ENDIF()
 
+set(BOOST_INCLUDE_DIR "" CACHE STRING "Path to boost include dir")
+if (NOT "${BOOST_INCLUDE_DIR}" STREQUAL "")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I ${BOOST_INCLUDE_DIR}")
+endif ()
+
 IF(NOT WITHOUT_SERVER)
   SET (MYSQLD_STATIC_PLUGIN_LIBS "" CACHE INTERNAL "")
   # Add storage engines and plugins.
