diff -u --recursive bundled_fbmysqlclient-orig/strings/ctype-simple.c bundled_fbmysqlclient/strings/ctype-simple.c
--- bundled_fbmysqlclient-orig/strings/ctype-simple.c	2021-07-09 15:48:32.000000000 -0500
+++ bundled_fbmysqlclient/strings/ctype-simple.c	2021-07-09 15:51:51.000000000 -0500
@@ -778,7 +778,7 @@
 {
   if (fast_integer_to_string)
   {
-    static_assert(sizeof(long int) == sizeof(long long),
+    _Static_assert(sizeof(long int) == sizeof(long long),
         "long int should be 64 bit");
     extern size_t u64toa_jeaiii_n(uint64_t n, char* b, const size_t len);
     extern size_t i64toa_jeaiii_n(int64_t i, char* b, const size_t len);
diff -u --recursive bundled_fbmysqlclient-orig/strings/int2str.c bundled_fbmysqlclient/strings/int2str.c
--- bundled_fbmysqlclient-orig/strings/int2str.c	2021-07-09 15:48:32.000000000 -0500
+++ bundled_fbmysqlclient/strings/int2str.c	2021-07-09 15:52:45.000000000 -0500
@@ -169,7 +169,7 @@
 {
   if (fast_integer_to_string)
   {
-    static_assert(sizeof(long int) == sizeof(long long),
+    _Static_assert(sizeof(long int) == sizeof(long long),
         "long int should be 64 bit");
     extern char* u64toa_jeaiii(uint64_t n, char* b);
     extern char* i64toa_jeaiii(int64_t i, char* b);
