Index: fbmysqlclient_external/libmysql/CMakeLists.txt
===================================================================
--- fbmysqlclient_external.orig/libmysql/CMakeLists.txt
+++ fbmysqlclient_external/libmysql/CMakeLists.txt
@@ -239,7 +239,7 @@ ELSE()
   LIST(APPEND LIBS_TO_LINK ${ZLIB_LIBRARY})
 ENDIF()
 
-IF(WITH_ZLIB STREQUAL "bundled")
+IF(WITH_ZSTD STREQUAL "bundled")
   LIST(APPEND LIBS_TO_MERGE ${ZSTD_LIBRARY})
 ELSE()
   LIST(APPEND LIBS_TO_LINK ${ZSTD_LIBRARY})
