diff --git a/CMakeLists.txt b/CMakeLists.txt
index 38bfe518e8c..26f4130828f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1022,11 +1022,21 @@ ENDIF()
 INCLUDE_DIRECTORIES(
   ${CMAKE_CURRENT_BINARY_DIR}
   ${CMAKE_CURRENT_BINARY_DIR}/include
-  ${CMAKE_SOURCE_DIR}
   ${CMAKE_SOURCE_DIR}/include
   ${CMAKE_CURRENT_BINARY_DIR}/libbinlogevents/include
   ${CMAKE_SOURCE_DIR}/libbinlogevents/export)
-
+IF(APPLE)
+  # `-iquote` is like `-I`, but *only* affects `#include "foo"`, not
+  # `#include <foo>`.
+  #
+  # This is needed as XCode 12.5/the MacOS 11.3 SDK contain
+  # `#include <version>` in several standard header files, which conflicts
+  # with the `VERSION` file in this repository, as MacOS filesystems are
+  # usually case-insensitive.
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -iquote ${CMAKE_SOURCE_DIR}")
+ELSE()
+  INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR})
+ENDIF()
 
 SET(WITH_SYSTEM_LIBS_DEFAULT OFF)
 
