Index: bundled_fbmysqlclient/cmake/zstd.cmake
===================================================================
--- bundled_fbmysqlclient.orig/cmake/zstd.cmake
+++ bundled_fbmysqlclient/cmake/zstd.cmake
@@ -31,6 +31,7 @@ MACRO (FIND_SYSTEM_ZSTD)
     CHECK_ZSTD_VERSION(${PATH_TO_ZSTD})
     SET(ZSTD_INCLUDE_DIR ${PATH_TO_ZSTD})
     SET(ZSTD_LIBRARY ${ZSTD_SYSTEM_LIBRARY})
+    INCLUDE_DIRECTORIES(BEFORE SYSTEM "${ZSTD_INCLUDE_DIR}")
     MESSAGE(STATUS "ZSTD_INCLUDE_DIR ${ZSTD_INCLUDE_DIR}")
     MESSAGE(STATUS "ZSTD_LIBRARY(system): ${ZSTD_LIBRARY}")
   ENDIF()
