diff --git a/fizz/CMakeLists.txt b/fizz/CMakeLists.txt
index 85b78981..5422ba9b 100644
--- a/fizz/CMakeLists.txt
+++ b/fizz/CMakeLists.txt
@@ -314,42 +314,42 @@ SET(FIZZ_TEST_INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX})
 if(BUILD_TESTS)
   find_package(GMock 1.8.0 MODULE REQUIRED)
   find_package(GTest 1.8.0 MODULE REQUIRED)
-endif()
 
-add_library(fizz_test_support
-  crypto/aead/test/TestUtil.cpp
-  crypto/test/TestUtil.cpp
-  protocol/ech/test/TestUtil.cpp
-  ${FIZZ_TEST_HEADERS}
-)
+  add_library(fizz_test_support
+    crypto/aead/test/TestUtil.cpp
+    crypto/test/TestUtil.cpp
+    protocol/ech/test/TestUtil.cpp
+    ${FIZZ_TEST_HEADERS}
+  )
 
-target_link_libraries(fizz_test_support
-  PUBLIC
-    fizz
-    ${LIBGMOCK_LIBRARIES}
-    ${GLOG_LIBRARY}
-)
+  target_link_libraries(fizz_test_support
+    PUBLIC
+      fizz
+      ${LIBGMOCK_LIBRARIES}
+      ${GLOG_LIBRARY}
+  )
 
-target_compile_definitions(fizz_test_support
-  PUBLIC
-    ${LIBGMOCK_DEFINES}
-)
+  target_compile_definitions(fizz_test_support
+    PUBLIC
+      ${LIBGMOCK_DEFINES}
+  )
 
-target_include_directories(fizz_test_support
-  SYSTEM
-  PUBLIC
-    ${LIBGMOCK_INCLUDE_DIR}
-    ${LIBGTEST_INCLUDE_DIRS}
-)
+  target_include_directories(fizz_test_support
+    SYSTEM
+    PUBLIC
+      ${LIBGMOCK_INCLUDE_DIR}
+      ${LIBGTEST_INCLUDE_DIRS}
+  )
 
-# export fizz headers and targets for unit tests utils
-# since other projects such as mvfst and proxygen use them
-install(
-  TARGETS fizz_test_support
-  EXPORT fizz-exports
-  ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
-  LIBRARY DESTINATION ${LIB_INSTALL_DIR}
-)
+  # export fizz headers and targets for unit tests utils
+  # since other projects such as mvfst and proxygen use them
+  install(
+    TARGETS fizz_test_support
+    EXPORT fizz-exports
+    ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
+    LIBRARY DESTINATION ${LIB_INSTALL_DIR}
+  )
+endif()
 
 foreach(dir ${FIZZ_TEST_HEADER_DIRS})
   get_filename_component(PARENT_DIR "/${dir}" DIRECTORY)
