diff --git a/balanced_brackets.php b/balanced_brackets.php
--- a/balanced_brackets.php
+++ b/balanced_brackets.php
@@ -3,16 +3,16 @@
 // https://rosettacode.org/wiki/Balanced_brackets#PHP

 function isbalanced(string $s): bool {
-  $balance = 0;
+  $bal = 0;
   for ($i = 0; $i < strlen($s); ++$i) {
     $ch = substr($s, $i, 1);
     if ($ch == '[') {
-      $balance++;
+      $bal++;
     } elseif ($ch == ']') {
-      $balance--;
+      $bal--;
     }
-    if ($balance < 0)
+    if ($bal < 0)
       return false;
   }
-  return ($balance == 0);
+  return ($bal == 0);
 }
diff --git a/fibonacci.php b/fibonacci.php
--- a/fibonacci.php
+++ b/fibonacci.php
@@ -1,9 +1,9 @@
 <?hh // strict

 function fibonacci($pos){
-  $a = array(0, 1);
+  $arr = array(0, 1);
   for ($i = 2; $i <= $pos; ++$i)
-    $a[$i] =
-      $a[$i-1] + $a[$i-2];
-  return $a[$pos];
+    $arr[$i] =
+      $arr[$i-1] + $arr[$i-2];
+  return $arr[$pos];
 }
