Index: bundled_folly/folly/io/async/ssl/OpenSSLUtils.cpp
===================================================================
--- bundled_folly.orig/folly/io/async/ssl/OpenSSLUtils.cpp
+++ bundled_folly/folly/io/async/ssl/OpenSSLUtils.cpp
@@ -387,7 +387,7 @@ std::vector<X509NameUniquePtr> OpenSSLUt
 
 std::vector<X509NameUniquePtr> OpenSSLUtils::subjectNamesInPEMBuffer(
     folly::ByteRange buffer) {
-  BioUniquePtr bio(BIO_new_mem_buf(buffer.data(), buffer.size()));
+  BioUniquePtr bio(BIO_new_mem_buf((void*)buffer.data(), buffer.size()));
   if (!bio) {
     throw std::runtime_error(
         "OpenSSLUtils::subjectNamesInPEMBuffer: failed to create BIO");
