Index: folly/CMake/folly-deps.cmake
===================================================================
--- folly.orig/CMake/folly-deps.cmake
+++ folly/CMake/folly-deps.cmake
@@ -126,11 +126,14 @@ if(ZSTD_FOUND)
   list(APPEND FOLLY_LINK_LIBRARIES ${ZSTD_LIBRARY})
 endif()
 
-find_package(Snappy MODULE)
-set(FOLLY_HAVE_LIBSNAPPY ${SNAPPY_FOUND})
-if (SNAPPY_FOUND)
-  list(APPEND FOLLY_INCLUDE_DIRECTORIES ${SNAPPY_INCLUDE_DIR})
-  list(APPEND FOLLY_LINK_LIBRARIES ${SNAPPY_LIBRARY})
+option(WITHOUT_LIBSNAPPY "Disable Snappy support" OFF)
+if (NOT WITHOUT_LIBSNAPPY)
+  find_package(Snappy MODULE)
+  set(FOLLY_HAVE_LIBSNAPPY ${SNAPPY_FOUND})
+  if (SNAPPY_FOUND)
+    list(APPEND FOLLY_INCLUDE_DIRECTORIES ${SNAPPY_INCLUDE_DIR})
+    list(APPEND FOLLY_LINK_LIBRARIES ${SNAPPY_LIBRARY})
+  endif()
 endif()
 
 find_package(LibDwarf)
