Index: folly/folly/CMakeLists.txt
===================================================================
--- folly.orig/folly/CMakeLists.txt
+++ folly/folly/CMakeLists.txt
@@ -21,7 +21,6 @@ target_link_libraries(follybenchmark PUB
 apply_folly_compile_options_to_target(follybenchmark)
 install(
   TARGETS follybenchmark
-  EXPORT folly
   RUNTIME DESTINATION ${BIN_INSTALL_DIR}
   LIBRARY DESTINATION ${LIB_INSTALL_DIR}
   ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
Index: folly/folly/debugging/exception_tracer/CMakeLists.txt
===================================================================
--- folly.orig/folly/debugging/exception_tracer/CMakeLists.txt
+++ folly/folly/debugging/exception_tracer/CMakeLists.txt
@@ -63,7 +63,6 @@ if (FOLLY_HAVE_ELF AND FOLLY_HAVE_DWARF)
       folly_exception_tracer_base
       folly_exception_tracer
       folly_exception_counter
-    EXPORT folly
     RUNTIME DESTINATION ${BIN_INSTALL_DIR}
     LIBRARY DESTINATION ${LIB_INSTALL_DIR}
     ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
