diff --git a/CMake/folly-deps.cmake b/CMake/folly-deps.cmake
index 92b8ce7e3..e0848785a 100644
--- a/CMake/folly-deps.cmake
+++ b/CMake/folly-deps.cmake
@@ -168,9 +168,15 @@ find_package(Backtrace)
 
 set(FOLLY_HAVE_BACKTRACE ${Backtrace_FOUND})
 set(FOLLY_HAVE_DWARF ${LIBDWARF_FOUND})
-if (NOT WIN32)
+if (UNIX AND NOT APPLE)
   set(FOLLY_USE_SYMBOLIZER ON)
 endif()
+if (NOT FOLLY_HAVE_ELF)
+  # Empty is falsey here, but ends up truthy in folly-config.h.cmake
+  set(FOLLY_HAVE_ELF OFF)
+  # The dwarf component depends on ELF
+  set(FOLLY_HAVE_DWARF OFF)
+endif()
 message(STATUS "Setting FOLLY_USE_SYMBOLIZER: ${FOLLY_USE_SYMBOLIZER}")
 message(STATUS "Setting FOLLY_HAVE_ELF: ${FOLLY_HAVE_ELF}")
 message(STATUS "Setting FOLLY_HAVE_DWARF: ${FOLLY_HAVE_DWARF}")
