diff --git a/thrift/lib/cpp2/CMakeLists.txt b/thrift/lib/cpp2/CMakeLists.txt
index f941688ab0..0f5ec47ae2 100644
--- a/thrift/lib/cpp2/CMakeLists.txt
+++ b/thrift/lib/cpp2/CMakeLists.txt
@@ -17,8 +17,8 @@ set(LIB_CPP2_HOME ${CMAKE_CURRENT_SOURCE_DIR})
 
 if (enable_tests)
   add_subdirectory(protocol/test)
+  add_subdirectory(test)
 endif ()
-add_subdirectory(test)
 
 #######
 # CMAKE variables only have local/subdirectory scope
