diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7b248a010..a6ad70eaf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -100,7 +100,10 @@ endif ()
 
 # Find required dependencies for thrift/lib
 if(lib_only OR build_all)
-  find_package(Gflags REQUIRED)
+  option(WITHOUT_GFLAGS "Do not use GFlags" OFF)
+  if(NOT WITHOUT_GFLAGS)
+    find_package(Gflags REQUIRED)
+  endif()
   find_package(Glog REQUIRED)
   find_package(folly CONFIG REQUIRED)
   find_package(fizz CONFIG REQUIRED)
